contracts/modules/SwapHub.sol

```diff
@@ -0,0 +1,182 @@
+
+        // Check liquidity
+        checkLiquidity(cache.accountIn);
+        if (cache.accountIn != cache.accountOut && assetStorageOut.users[cache.accountOut].owed != 0)
+            checkLiquidity(cache.accountOut);
+    }
+
+    /// @notice Repay debt by selling another deposited token
+    /// @param subAccountIdIn sub-account holding the sold token. 0 for primary, 1-255 for a sub-account
+    /// @param subAccountIdOut sub-account to receive the bought token. 0 for primary, 1-255 for a sub-account
+    /// @param swapHandler address of a swap handler to use
+    /// @param params struct defining the requested trade
+    /// @param targetDebt how much debt should remain after calling the function
+    function swapAndRepay(uint subAccountIdIn, uint subAccountIdOut, address swapHandler, ISwapHandler.SwapParams memory params, uint targetDebt) external nonReentrant {
+        SwapCache memory cache = initSwap(subAccountIdIn, subAccountIdOut, params);
+
+        emit RequestSwapHubRepay(
+            cache.accountIn,
+            cache.accountOut,
+            params.underlyingIn,
+            params.underlyingOut,
+            targetDebt,
+            swapHandler
+        );
+
+        // Adjust params for repay
+        require(params.mode == 1, "e/swap-hub/repay-mode");
+        uint owed = getCurrentOwed(eTokenLookup[cache.eTokenOut], cache.assetCacheOut, cache.accountOut) / cache.assetCacheOut.underlyingDecimalsScaler;
+        require (owed > targetDebt, "e/swap-hub/target-debt");
+        params.amountOut = owed - targetDebt;
+
+        uint amountOut = swapInternal(cache, swapHandler, params);
+
+        // Process repay
+        AssetStorage storage assetStorageOut = eTokenLookup[cache.eTokenOut];
+        cache.assetCacheOut.poolSize = decodeExternalAmount(cache.assetCacheOut, cache.preBalanceOut + amountOut);
+        decreaseBorrow(assetStorageOut, cache.assetCacheOut, assetStorageOut.dTokenAddress, cache.accountOut, decodeExternalAmount(cache.assetCacheOut, amountOut));
+        logAssetStatus(cache.assetCacheOut);
+
+        // Check liquidity only for outgoing account, repay can't lower the health score or cause borrow isolation error
+        checkLiquidity(cache.accountIn);
+    }
+
+    function swapInternal(SwapCache memory cache, address swapHandler, ISwapHandler.SwapParams memory params) private returns (uint) {
+        // Adjust requested amount in
+        (uint amountInScaled, uint amountInInternal) = withdrawAmounts(eTokenLookup[cache.eTokenIn], cache.assetCacheIn, cache.accountIn, params.amountIn);
+        require(cache.assetCacheIn.poolSize >= amountInScaled, "e/swap-hub/insufficient-pool-size");
+        params.amountIn = amountInScaled / cache.assetCacheIn.underlyingDecimalsScaler;
+
+        // Supply handler, for exact output amount transfered serves as an implicit amount in max.
+        Utils.safeTransfer(cache.assetCacheIn.underlying, swapHandler, params.amountIn);
```

Comment:
may be simplified to:
`Utils.safeTransfer(params.underlyingIn, swapHandler, params.amountIn);`

Reply:
good spot